packages/algorand-js/src/runtime/runtime.ts

```diff
@@ -0,0 +1,226 @@
+        txns.push(encodedTxnObj);
+      }
+      assignGroupID(txns); // assign unique groupID to all transactions in the array/group
+      return [txns[0], txns]; // by default current txn is the first txn (hence txns[0])
+    } else {
+      // if not array, then create a single transaction
+      const mockParams = mockSuggestedParams(txnParams.payFlags);
+      const tx = mkTransaction(txnParams, mockParams);
+
+      const encodedTxnObj = tx.get_obj_for_encoding() as Txn;
+      encodedTxnObj.txID = tx.txID();
+      return [encodedTxnObj, [encodedTxnObj]];
+    }
+  }
+
+  // updates account balance as per transaction parameters
+  updateBalance (txnParam: ExecParams, account: StoreAccount): void {
+    switch (txnParam.type) {
+      case TransactionType.TransferAlgo: {
+        switch (account.address) {
+          case txnParam.fromAccount.addr: {
+            account.amount -= txnParam.amountMicroAlgos; // remove 'x' algo from sender
+            break;
+          }
+          case txnParam.toAccountAddr: {
+            account.amount += txnParam.amountMicroAlgos; // add 'x' algo to receiver
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Description: update current state and account balances
+   * @param txnParams : Transaction parameters
+   * @param accounts : accounts passed by user
+   */
+  prepareFinalState (txnParams: ExecParams | ExecParams[], accounts: StoreAccount[]): void {
+    if (Array.isArray(txnParams)) { // if txn is a group, update balance as per 'each' transaction
+      for (const txnParam of txnParams) {
+        for (const acc of accounts) {
+          this.updateBalance(txnParam, acc);
+        }
+      }
+    } else {
+      // for a single (stand alone) transaction
+      for (const acc of accounts) {
+        this.updateBalance(txnParams, acc);
+      }
+    }
+  }
+
+  /**
+   * Description: this function executes a transaction based on a smart
+   * contract logic and updates state afterwards
+   * @param txnParams : Transaction parameters
+   * @param fileName : smart contract file (.teal) name in assets/
+   * @param args : external arguments to smart contract
+   */
+  async executeTx (txnParams: ExecParams | ExecParams[], fileName: string,
+    args: Uint8Array[]): Promise<void> {
+    const [tx, gtxs] = this.createTxnContext(txnParams); // get current txn and txn group (as encoded obj)
+
+    // initialize context before each execution
+    this.ctx = {
+      state: _.cloneDeep(this.store), // state is a deep copy of store
+      tx: tx,
+      gtxs: gtxs,
+      args: args
+    };
+
+    const program = getProgram(fileName); // get TEAL code as string
+    const interpreter = new Interpreter();
+    await interpreter.execute(program, this);
+
+    this.store = this.ctx.state; // update state after successful execution('local-state', 'global-state'..)
+    this.prepareFinalState(txnParams, this.accounts); // update account balances
```

Comment:
Why do we have accounts both in `this.accounts` and `this.store.accounts`? Please update if this is a mistake. Probably we should only have `this.store.accounts`

Reply:
`this.accounts` was kept to update the accounts passed by the user (in runtime), along with accountMap. So that it's easier for the end user to check the updates in original account (rather then getting from map).
Reverting.